### PR TITLE
Use a standard list item component for all content types

### PR DESCRIPTION
**Pull request by geoffreylitt**
Still WIP, looking for high level feedback before going any further @pvh @paulsonnentag 

---

This change is coming out of discussion on #15 about avoiding bloating the API surface of `<Content />`.

The general idea is to use a single default list item component to render all content types within a content list. This removes duplication across all the various `<X>InList` components and standardizes the UI.

(BTW, we should probably still allow defining a custom component if there's really a good reason, but I think that might be better thought of as "extending the default" with some future extension mechanism, than "defining an entirely custom list item component")

I'm also experimenting w/ isolating the mapping logic for turning the _data_ of a given document into a list item--i.e., figure out what the title is, what the unread badge should do, etc. Do not tightly couple this logic with either the content type or with the list item view. This approximates the value of Cambria: content types don't need to know that there is this "list item" context where their data might be rendered, and vice versa. It should be easy to add another content type without worrying about all the possible data contexts out there, or to add a new renderer of list items.

---

- [x] roughly sketch the direction for feedback
- [x] Shouldn't need to explicitly specify `DefaultInList` within each content type -- for some reason the default component fallback isn't working
- [x] Computation of unread counts is broken somewhere in the chain, fix
- [x] Apply this pattern to all content types
- [x] Clean up and remove newly unnecessary XInList components

### CODE REVIEW

**Comment by paulsonnentag** [comment]
I like the idea. I think it's good if we can define new interfaces like the title without always having to extend the ContentType definition. I think creating new interfaces is fine as long as they have a fallback that covers all possible documents. Have you thought about how a content type could provide a custom implementation? For example that the first line of text is used as a fallback if no explicit title is set. I think this behavior should be a custom implementation in the text content. I don't think this is something we have to solve right now, but eventually, we should have a more principled solution.

**Comment by geoffreylitt** [comment]
This PR is ready to merge from my perspective, would be great to get a review @pvh @paulsonnentag. If it looks good, next steps will be:

- rebase against #16 and merge
- Then follow up in a new PR adding back the context menu change, now that we have a standardized UI

I've changed Text, Thread, Board, and ContentList to all use the default renderer in the `list` context now. The default renderer uses `src/lenses/ListItem.ts`, which defines a _data_ mapping (not a specific UI view) from those content types to a list item.

I've left custom renderers in place for now for files, images, and workspaces—it seemed like those were different enough to possibly justify a custom renderer.

> Have you thought about how a content type could provide a custom implementation? For example that the first line of text is used as a fallback if no explicit title is set.

Yes, @paulsonnentag see `ListItem.ts` where we define custom implementations of the data mapping which branch on content types. I intentionally located this logic outside of the content type itself. The reasoning: we will grow to have many content types and many "data schemas" which we'd like to convert them to. It shouldn't be necessary for every content type to define a ton of different mappings to all possible data schemas; instead there should be a library of lenses which manages this. This is intended as a little step towards this direction.

One related question though which we do need to figure out: for now, `hasUnseenChanges` still lives on content types. But maybe we don't like this? (see https://github.com/pvh/blutack/blob/6cf0df09ffcb09904017d192d05807e0adced2bc/src/components/pushpin-code/ContentTypes.ts#L47-L48)

Another way to handle `hasUnseenChanges` would be to split out another lens file that maps documents to an "unseen changes" schema. So maybe we could have a bunch of lenses mapping each doc to "title" and "unseen changes" schemas, which are both used in various other places. The metaphor is, "I want to open doc X (of arbitrary type) through the schema HasUnseenChanges"

File: src/components/content-types/TextContent.tsx [hunk 1]
```diff
@@ -33,6 +33,7 @@ import {
 import { createDocumentLink } from "../pushpin-code/Url"
 import memoize from "lodash.memoize"
 import { Doc, getHeads } from "@automerge/automerge"
+import DefaultInList from "./defaults/DefaultInList"
```
Comment (pvh) [inline src/components/content-types/TextContent.tsx]: looks like you can cut this?

**Comment by geoffreylitt** [comment]
I've made some changes in response to some new ideas from @pvh in a conversation today.

The goal is now to remove the "list" context and replace it with two smaller contexts: "title" and "badge". This change has some immediate practical benefits, but also is taking a step towards a bigger rethink of contexts and content types.

The title component is pretty standardized--that is, we don't really have any content types that need to override the UI in fancy ways. All we need is a way to fetch the appropriate data from the document. So, we have a `DefaultInTitle` component that serves as the default renderer in the title context. All it does is read the doc thru a lens using a "HasTitle" schema, and then passes the data down to the standard `TitleWithSubtitle`:

```
const doc = readWithSchema({
    doc: rawDoc,
    type,
    lastSeenHeads,
    schema: "HasTitle",
  }) as HasTitle

  return (
    <TitleWithSubtitle
      title={doc.title}
      titleEditorField={doc.titleEditorField ?? ""}
      subtitle={doc.subtitle}
      documentId={documentId}
      editable={doc.titleEditorField !== null}
    />
  )
```

We take a similar approach with badges. We have a default component, and usually we just read arbitrary docs thru a lens. However, badges are a little more specialized, we have some cases like Contact where we do really want a special custom UI. So, for Contacts we override the default component and render a custom Badge. But, we still use the standard title UI for contacts, because we don't need to override that part.

---

The more general philosophy emerging here is:

- More granular contexts. In fact, these aren't really "contexts" anymore, I would say they are actually closer to "renderers". That is, we are rendering a doc "as a title", not "in a list".  Maybe we can get rid of all contexts and replace with renderers? I think this makes more sense generally, to define how to render a doc "as an X" rather than "inside a Y"
- Use data lenses wherever possible. This is good because 1) it preserves more standardized UI without unnecessary variation, 2) it's much easier to define new renderers if the data lensing logic is decoupled from a particular view
- If you need to override UI, do it at the smallest possible granularity

I think broadly you can see a shape emerging that's something like:

- A document can contain data in 1 or more schemas
- You can open a doc in any schema, and it'll be opened thru lenses
- There's a default way to render any doc in any schema, but you can also override with custom renderers. (There's a missing piece here that's still fuzzy, need to think more about where the layers of indirection are in the system...)

---

Current state of this PR: I believe it's mostly working. todos:

- I've moved contact and file over to the new system, still a few more content types to consider
- There's still a "list" context which is just a passthru to badge + title, maybe we want to get rid of that? Or maybe fine to keep it, need to consider that more.
- general cleanup

Hoping to land it tomorrow before my vacation, maybe doable as long as we don't pull in too much more scope in this whole philosophical refactor...

Curious if this direction seems good @pvh @paulsonnentag

**Comment by pvh** [comment]
Okay, so I took a good look into this and I really like how it does away with so many of the contexts that are out there! 

I went one step further and decomposed the "list" into badge/title for good measure which let me move the name up into the message group on the `lenses-without-list` branch. I'm not sure if we want to go that way or not but it exists. I *do* think we might want a generic ListItem presentational component that at least tidies up these queries.

I also fixed a long-standing weirdness which was how the ContentTypes register themselves. Now they export a ContentType object like { threadContentType } which gets passed to ContentTypes.register() explicitly. I think it's better to expose the control flow but it's a little ugly to read. Improvements welcome.

Here's what I would do from here (and known issues):

1) Get rid of the title context. Let's just use the summary and the developer can choose to render as a Label or an Editor depending on the context. The benefit here is that we don't nest an extra useDocument in here and burn more frames. This will also help with a rendering issue I noticed: the Title component has margins on it so the names are strangely positioned in the chat log. A little styling would go a long way here.

2) File content can take a long time to load and in the meantime, we get no badge at all. We should return a stand-in badge with a generic icon while we wait.

3) Keyboard shortcuts for the Omnibox have stopped working. I do use them all the time and it's new so I'm not sure what happened here. It's probably an easy fix (I hope it wasn't me but I have not had time to check).

That's all I can think of! It's *so* slow right now. Not sure why. Could be unrelated but worth checking that as well.

**Pull request merged**